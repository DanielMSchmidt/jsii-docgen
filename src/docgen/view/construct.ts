import * as reflect from 'jsii-reflect';
import { Markdown } from '../render/markdown';
import { ConstructSchema } from '../schema';
import { Transpile, TranspiledType } from '../transpile/transpile';
import { Class } from './class';

export class Construct {
  private readonly construct: Class;
  constructor(transpile: Transpile, klass: reflect.ClassType, linkFormatter: (type: TranspiledType) => string) {
    this.construct = new Class(transpile, klass, linkFormatter);
  }
  public toMarkdown(): Markdown {
    return this.construct.toMarkdown();
  }

  public toJson(): ConstructSchema {
    return this.construct.toJson();
  }
}
