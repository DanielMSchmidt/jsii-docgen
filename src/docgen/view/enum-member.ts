import * as reflect from 'jsii-reflect';
import { Markdown } from '../render/markdown';
import { EnumMemberSchema } from '../schema';
import { Transpile, TranspiledEnumMember } from '../transpile/transpile';

export class EnumMember {
  private readonly transpiled: TranspiledEnumMember;
  constructor(transpile: Transpile, private readonly em: reflect.EnumMember) {
    this.transpiled = transpile.enumMember(em);
  }
  public toMarkdown(): Markdown {
    const md = new Markdown({
      id: `${this.transpiled.fqn}`,
      header: {
        title: this.transpiled.name,
        pre: true,
        strike: this.em.docs.deprecated,
      },
    });

    if (this.em.docs.deprecated) {
      md.bullet(
        `${Markdown.italic('Deprecated:')} ${this.em.docs.deprecationReason}`,
      );
      md.lines('');
    }

    if (this.em.docs) {
      md.docs(this.em.docs);
    }

    md.split();
    md.lines('');

    return md;
  }

  public toJson(): EnumMemberSchema {
    return {
      id: this.transpiled.fqn,
      name: this.transpiled.name,
      deprecated: this.em.docs.deprecated,
      deprecationReason: this.em.docs.deprecationReason,
      docs: this.em.docs.toString(),
    };
  }
}
