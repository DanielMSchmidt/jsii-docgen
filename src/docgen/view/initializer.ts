import * as reflect from 'jsii-reflect';
import { Markdown } from '../render/markdown';
import { InitializerSchema } from '../schema';
import { Transpile, TranspiledCallable, TranspiledType } from '../transpile/transpile';
import { Parameter } from './parameter';

export class Initializer {
  private readonly transpiled: TranspiledCallable;
  private readonly parameters: Parameter[];
  constructor(
    private readonly transpile: Transpile,
    initializer: reflect.Initializer,
  ) {
    this.transpiled = transpile.callable(initializer);
    this.parameters = this.transpiled.parameters.map(
      (p) => new Parameter(this.transpile, p),
    );
  }

  public toMarkdown(
    linkFormatter: (type: TranspiledType) => string,
  ): Markdown {
    const md = new Markdown({
      id: `${this.transpiled.parentType.fqn}.Initializer`,
      header: {
        title: 'Initializers',
      },
    });

    md.code(
      this.transpile.language.toString(),
      `${this.transpiled.import}`,
      '',
      ...this.transpiled.invocations,
    );

    for (const parameter of this.parameters) {
      md.section(parameter.toMarkdown(linkFormatter));
    }

    return md;
  }

  public toJson(): InitializerSchema {
    const md = new Markdown();
    md.code(
      this.transpile.language.toString(),
      `${this.transpiled.import}`,
      '',
      ...this.transpiled.invocations,
    );

    return {
      id: `${this.transpiled.parentType.fqn}.Initializer`,
      snippet: md.render(),
      parameters: this.parameters.map((parameter) => parameter.toJson()),
    };
  }
}
