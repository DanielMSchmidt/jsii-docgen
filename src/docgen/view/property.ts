import * as reflect from 'jsii-reflect';
import { Markdown } from '../render/markdown';
import { PropertySchema } from '../schema';
import { Transpile, TranspiledProperty, TranspiledType } from '../transpile/transpile';

export class Property {
  private readonly transpiled: TranspiledProperty;
  constructor(
    private readonly transpile: Transpile,
    private readonly property: reflect.Property,
  ) {
    this.transpiled = transpile.property(property);
  }

  public toMarkdown(linkFormatter: (type: TranspiledType) => string): Markdown {
    const optionality = this.property.const
      ? undefined
      : this.property.optional
        ? 'Optional'
        : 'Required';

    const md = new Markdown({
      id: `${this.transpiled.parentType.fqn}.property.${this.transpiled.name}`,
      header: {
        title: this.transpiled.name,
        sup: optionality,
        pre: true,
        strike: this.property.docs.deprecated,
      },
    });

    if (this.property.docs.deprecated) {
      md.bullet(
        `${Markdown.italic('Deprecated:')} ${
          this.property.docs.deprecationReason
        }`,
      );
      md.lines('');
    }

    if (!this.property.const) {
      md.code(this.transpile.language.toString(), this.transpiled.declaration);
    }

    const metadata: any = {
      Type: this.transpiled.typeReference.toString({
        typeFormatter: (t) => `[${Markdown.pre(t.fqn)}](${linkFormatter(t)})`,
        stringFormatter: Markdown.pre,
      }),
    };

    if (this.property.spec.docs?.default) {
      metadata.Default = Markdown.sanitize(this.property.spec.docs?.default);
    }

    for (const [key, value] of Object.entries(metadata)) {
      md.bullet(`${Markdown.italic(`${key}:`)} ${value}`);
    }
    md.lines('');

    if (this.property.docs) {
      md.docs(this.property.docs);
    }

    md.split();

    return md;
  }

  public toJson(): PropertySchema {
    return {
      id: `${this.transpiled.parentType.fqn}.${this.transpiled.name}`,
      name: this.transpiled.name,
      optional: this.property.const ? false : this.property.optional,
      deprecated: this.property.docs.deprecated,
      deprecationReason: this.property.docs.deprecationReason,
      docs: this.property.docs.toString(),
      default: this.property.spec.docs?.default,
      type: this.transpiled.typeReference.toJson(),
    };
  }
}
