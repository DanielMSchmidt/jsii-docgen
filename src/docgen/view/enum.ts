import * as reflect from 'jsii-reflect';
import { Markdown } from '../render/markdown';
import { EnumSchema } from '../schema';
import { Transpile } from '../transpile/transpile';
import { EnumMember } from './enum-member';

export class Enum {
  private readonly members: EnumMember[];
  constructor(
    private readonly transpile: Transpile,
    private readonly enu: reflect.EnumType,
  ) {
    this.members = enu.members.map((em) => new EnumMember(transpile, em));
  }

  public toMarkdown(): Markdown {
    const transpiled = this.transpile.enum(this.enu);
    const md = new Markdown({ header: { title: transpiled.name } });

    if (this.enu.docs) {
      md.docs(this.enu.docs);
    }

    for (const m of this.members) {
      md.section(m.toMarkdown());
    }

    return md;
  }

  public toJson(): EnumSchema {
    const transpiled = this.transpile.enum(this.enu);
    return {
      name: transpiled.name,
      docs: this.enu.docs.toString(),
      members: this.members.map((member) => member.toJson()),
    };
  }
}
