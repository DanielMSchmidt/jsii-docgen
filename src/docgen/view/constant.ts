import * as reflect from 'jsii-reflect';
import { Markdown } from '../render/markdown';
import { PropertySchema } from '../schema';
import { Transpile, TranspiledType } from '../transpile/transpile';
import { Property } from './property';

export class Constant {
  private readonly constant: Property;
  constructor(transpile: Transpile, property: reflect.Property, linkFormatter: (type: TranspiledType) => string) {
    this.constant = new Property(transpile, property, linkFormatter);
  }
  public toMarkdown(): Markdown {
    return this.constant.toMarkdown();
  }

  public toJson(): PropertySchema {
    return this.constant.toJson();
  }
}
