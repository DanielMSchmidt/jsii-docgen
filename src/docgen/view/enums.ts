import * as reflect from 'jsii-reflect';
import { Markdown } from '../render/markdown';
import { EnumSchema } from '../schema';
import { Transpile } from '../transpile/transpile';
import { Enum } from './enum';

export class Enums {
  private readonly enums: Enum[];
  constructor(transpile: Transpile, enums: reflect.EnumType[]) {
    this.enums = enums.map((e) => new Enum(transpile, e));
  }

  public toMarkdown(): Markdown {
    if (this.enums.length === 0) {
      return Markdown.EMPTY;
    }

    const md = new Markdown({ header: { title: 'Enums' } });
    for (const e of this.enums) {
      md.section(e.toMarkdown());
    }
    return md;
  }

  public toJson(): EnumSchema[] {
    return this.enums.map((e) => e.toJson());
  }
}
