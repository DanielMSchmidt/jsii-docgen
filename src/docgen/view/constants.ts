import * as reflect from 'jsii-reflect';
import { Markdown } from '../render/markdown';
import { PropertySchema } from '../schema';
import { Transpile, TranspiledType } from '../transpile/transpile';
import { Constant } from './constant';

export class Constants {
  private readonly constants: Constant[];
  constructor(transpile: Transpile, properties: reflect.Property[]) {
    this.constants = properties
      .filter((p) => !p.protected && p.const)
      .map((p) => new Constant(transpile, p));
  }

  public toMarkdown(linkFormatter: (type: TranspiledType) => string): Markdown {
    if (this.constants.length === 0) {
      return Markdown.EMPTY;
    }

    const md = new Markdown({ header: { title: 'Constants' } });
    for (const c of this.constants) {
      md.section(c.toMarkdown(linkFormatter));
    }
    return md;
  }

  public toJson(): PropertySchema[] {
    return this.constants.map((constant) => constant.toJson());
  }
}
