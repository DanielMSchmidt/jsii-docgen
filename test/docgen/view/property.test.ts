import * as reflect from 'jsii-reflect';
import { CSharpTranspile } from '../../../src/docgen/transpile/csharp';
import { JavaTranspile } from '../../../src/docgen/transpile/java';
import { PythonTranspile } from '../../../src/docgen/transpile/python';
import { TranspiledType } from '../../../src/docgen/transpile/transpile';
import { TypeScriptTranspile } from '../../../src/docgen/transpile/typescript';
import { Property } from '../../../src/docgen/view/property';
import { Assemblies } from '../assemblies';

const assembly: reflect.Assembly = Assemblies.instance.withoutSubmodules;

describe('python', () => {
  const transpile = new PythonTranspile();
  test('markdown snapshot', () => {
    const parameter = new Property(
      transpile,
      assembly.system.interfaces[0].allProperties[0],
    );
    expect(parameter.toMarkdown((t: TranspiledType) => `#${t.fqn}`).render()).toMatchSnapshot();
  });

  test('json snapshot', () => {
    const parameter = new Property(
      transpile,
      assembly.system.interfaces[0].allProperties[0],
    );
    expect(parameter.toJson()).toMatchSnapshot();
  });
});

describe('typescript', () => {
  const transpile = new TypeScriptTranspile();
  test('markdown snapshot', () => {
    const parameter = new Property(
      transpile,
      assembly.system.interfaces[0].allProperties[0],
    );
    expect(parameter.toMarkdown((t: TranspiledType) => `#${t.fqn}`).render()).toMatchSnapshot();
  });

  test('json snapshot', () => {
    const parameter = new Property(
      transpile,
      assembly.system.interfaces[0].allProperties[0],
    );
    expect(parameter.toJson()).toMatchSnapshot();
  });
});

describe('java', () => {
  const transpile = new JavaTranspile();
  test('markdown snapshot', () => {
    const parameter = new Property(
      transpile,
      assembly.system.interfaces[0].allProperties[0],
    );
    expect(parameter.toMarkdown((t: TranspiledType) => `#${t.fqn}`).render()).toMatchSnapshot();
  });

  test('json snapshot', () => {
    const parameter = new Property(
      transpile,
      assembly.system.interfaces[0].allProperties[0],
    );
    expect(parameter.toJson()).toMatchSnapshot();
  });
});

describe('csharp', () => {
  const transpile = new CSharpTranspile();
  test('markdown snapshot', () => {
    const parameter = new Property(
      transpile,
      assembly.system.interfaces[0].allProperties[0],
    );
    expect(parameter.toMarkdown((t: TranspiledType) => `#${t.fqn}`).render()).toMatchSnapshot();
  });

  test('json snapshot', () => {
    const parameter = new Property(
      transpile,
      assembly.system.interfaces[0].allProperties[0],
    );
    expect(parameter.toJson()).toMatchSnapshot();
  });
});
