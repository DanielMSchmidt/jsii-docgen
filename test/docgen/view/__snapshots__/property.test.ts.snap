// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`csharp json snapshot 1`] = `
Object {
  "default": undefined,
  "deprecated": false,
  "deprecationReason": undefined,
  "docs": "\`AWS::ECR::PublicRepository.RepositoryCatalogData\`.",
  "id": "Amazon.CDK.AWS.ECR.CfnPublicRepositoryProps.RepositoryCatalogData",
  "name": "RepositoryCatalogData",
  "optional": true,
  "type": Object {
    "name": "object",
  },
}
`;

exports[`csharp markdown snapshot 1`] = `
" \`RepositoryCatalogData\`<sup>Optional</sup> <a name=\\"Amazon.CDK.AWS.ECR.CfnPublicRepositoryProps.property.RepositoryCatalogData\\"></a>

\`\`\`csharp
public object RepositoryCatalogData { get; set; }
\`\`\`

- *Type:* \`object\`

\`AWS::ECR::PublicRepository.RepositoryCatalogData\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorycatalogdata](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorycatalogdata)

---
"
`;

exports[`java json snapshot 1`] = `
Object {
  "default": undefined,
  "deprecated": false,
  "deprecationReason": undefined,
  "docs": "\`AWS::ECR::PublicRepository.RepositoryCatalogData\`.",
  "id": "software.amazon.awscdk.services.ecr.CfnPublicRepositoryProps.repositoryCatalogData",
  "name": "repositoryCatalogData",
  "optional": true,
  "type": Object {
    "name": "java.lang.Object",
  },
}
`;

exports[`java markdown snapshot 1`] = `
" \`repositoryCatalogData\`<sup>Optional</sup> <a name=\\"software.amazon.awscdk.services.ecr.CfnPublicRepositoryProps.property.repositoryCatalogData\\"></a>

\`\`\`java
public java.lang.Object getRepositoryCatalogData();
\`\`\`

- *Type:* \`java.lang.Object\`

\`AWS::ECR::PublicRepository.RepositoryCatalogData\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorycatalogdata](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorycatalogdata)

---
"
`;

exports[`python json snapshot 1`] = `
Object {
  "default": undefined,
  "deprecated": false,
  "deprecationReason": undefined,
  "docs": "\`AWS::ECR::PublicRepository.RepositoryCatalogData\`.",
  "id": "aws_cdk.aws_ecr.CfnPublicRepositoryProps.repository_catalog_data",
  "name": "repository_catalog_data",
  "optional": true,
  "type": Object {
    "name": "typing.Any",
  },
}
`;

exports[`python markdown snapshot 1`] = `
" \`repository_catalog_data\`<sup>Optional</sup> <a name=\\"aws_cdk.aws_ecr.CfnPublicRepositoryProps.property.repository_catalog_data\\"></a>

\`\`\`python
repository_catalog_data: typing.Any
\`\`\`

- *Type:* \`typing.Any\`

\`AWS::ECR::PublicRepository.RepositoryCatalogData\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorycatalogdata](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorycatalogdata)

---
"
`;

exports[`typescript json snapshot 1`] = `
Object {
  "default": undefined,
  "deprecated": false,
  "deprecationReason": undefined,
  "docs": "\`AWS::ECR::PublicRepository.RepositoryCatalogData\`.",
  "id": "@aws-cdk/aws-ecr.CfnPublicRepositoryProps.repositoryCatalogData",
  "name": "repositoryCatalogData",
  "optional": true,
  "type": Object {
    "name": "any",
  },
}
`;

exports[`typescript markdown snapshot 1`] = `
" \`repositoryCatalogData\`<sup>Optional</sup> <a name=\\"@aws-cdk/aws-ecr.CfnPublicRepositoryProps.property.repositoryCatalogData\\"></a>

\`\`\`typescript
public readonly repositoryCatalogData: any;
\`\`\`

- *Type:* \`any\`

\`AWS::ECR::PublicRepository.RepositoryCatalogData\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorycatalogdata](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorycatalogdata)

---
"
`;
